docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+
+> **NOTE** \
+Module fields are declared using the `field_name: TypeHint` syntax (same as dataclasses). Without a type hint, an attribute is considered a static property of the class. In case you cannot specify the type you can use `typing.Any` as a wildcard type.
+
+
+## Compact Modules
+
+Linen provides an alternative API for defining modules more compactly. This is especially useful for the common case where the Module consists of only one method that uses parameters and/or sub-modules. Using the compact API the MLP can be rewritten as follows:
+
+
+```python
+class CompactMLP(nn.Module):
+  hidden_size: int
+  out_size: int
+
+  @nn.compact
+  def __call__(self, x):
+    a = nn.Dense(self.hidden_size)(x)
+    h = nn.relu(a)
+    return nn.Dense(self.out_size)(h)
+
+```
+
+
+A compact `Module` is similar in spirit to a function. It offers a concise notation and restricts external interaction to the inputs and return values of the function. In this case the concise notation might make it easier for others to understand what the Module is doing. There is no need to jump back and forth between the setup and call function to understand what the submodules are doing. Instead, simply reading the call method from top to bottom once should give a clear picture of what the Module is doing. This can make a significant difference if you are implementing complex Modules with many hyper parameters.
+
+Another benefit of defining submodules and/or variables inline is that you can use arguments to your method when constructing variables. The most common example of this is using shape information to determine the shape of a parameter like this:
+
+
+```python
+class CompactScaledMLP(nn.Module):
+  hidden_size: int
+  out_size: int
+
+  @nn.compact
+  def __call__(self, x):
+    scale = self.param("scale", x.shape[-1:], nn.initializers.ones)
+    x *= scale
+    a = nn.Dense(self.hidden_size)(x)
+    h = nn.relu(a)
+    return nn.Dense(self.out_size)(h)
+```
+
+
+Many of the standard linen layers like `nn.Dense` use this type of shape inference already to avoid the need to specify input shapes (like the number of input features to a Dense layer).
+
+### Compact control flow
+
+The order in which you define submodules determines the name of a submodule if none is provided explicitly (using the name= construction argument). Because the name determines how parameters are mapped to submodules, you must be careful about mixing control flow with auto-generated names. Using control flow can change the order or remove certain submodules altogether. This is useful in the case where certain submodules should only exist depending on some construction argument. However, when control flow depends on the input arguments to the Module you should be careful. For example the following Module will break:
```

Comment:
```suggestion
The order in which you define submodules determines the name of a submodule if none is provided explicitly (using the `name=` keyword argument pass to the module's constructor). Because the name determines how parameters are mapped to submodules, you must be careful about mixing control flow with auto-generated names. Using control flow can change the order or remove certain submodules altogether. This is useful in the case where certain submodules should only exist depending on some construction argument. However, when control flow depends on the input arguments to the Module you should be careful. For example the following Module will break:
```

Reply:
@marcvanzee your version + some formatting

```suggestion
The order in which you define submodules determines the name of a submodule if `none` is provided explicitly (using the `name=` keyword argument pass to the module's constructor). Because the `name` determines how parameters are mapped to submodules, you must be careful about mixing control flow with auto-generated names. Using control flow can change the order or remove certain submodules altogether. This is useful in cases where certain submodules should only exist depending on some construction argument. However, when control flow depends on the input arguments to the Module, you should be careful. For example, the following Module will break:
```